### ISSUE TITLE
Model Data Isolation

### ISSUE THREAD

**Issue by Souheil-Yazji** [StatCan/openmpp#31]
According to OpenM inc. the model data should be isolated between users by design. We may want to reconsider using shared blob storage for this. 

Investigate if the following is possible and whether it will resolve the issue.

Modify the Run_MPIJobs script to:
- make a copy of the models database
- run the model on the copied and isolated database
- merge the results back into the primary database file.

## Final Update Summary 
There were several issues with the data storage layer of OpenM when integrating with AAW, these are documented in this issue. 

I had initially thought this was due to concurrency, leading to race conditions on writes to the SQLite file and therefore "corrupting" the database or encountering the `transaction is already active`/`IO error`. 

The next step was to isolate model data and to safeguard it from concurrent rights through MPI. This was achieved by testing non-mpi oncosim model run behavior, which still yielded issues. I also consulted OpenM on how model runs and MPI interact with the database and they were adamant there there's no concurrent writes to the database and that only one single `node` was responsible for writing to the database. They also have MPI working fine on GCP, and I've also tested MPI runs of oncosim on my WSL instance, and that yielded no corruption. 

With these results in mind, possibility of concurrency being the issue is ruled out. The next layer to investigate is the data storage medium. AAW uses blob storage, which has limitation in terms of I/O since it uses a REST API to interact with blobs. We can see in this [SQLite doc](https://sqlite.org/cloudsqlite/doc/trunk/www/index.wiki) details about cloud specific sqlite, intended for storage in blob storage such as Azure Blob Storage, which stores data as blob blocks and not in the conventional sqlite file format. 

As a work around, Azure File share was used, which offers improved I/O and a directory structure more in line with conventional NFS, and also offers SMB/NFS protocol storage. This seemed to resolve the issues with database writes. 

I'll be marking this as complete and creating a new issue for the Azure file share implementation.

**Comment by KrisWilliamson** [comment]
First issue, it is not recommended to simply make a copy of a sqlite database as it is possible that other processes are using it, which could lead to possible corruption.

Instead it is recommended to use the sqlite  `.backup FILE` or `.save FILE` functions.   
This can be invoked from the command line like `>sqlite3 my_db.sqlite ".backup backup_db.sqlite"`

**Comment by KrisWilliamson** [comment]
Merging two sqlite database is more difficult as there may be re-use of keys.

**Comment by Souheil-Yazji** [comment]
> First issue, it is not recommended to simply make a copy of a sqlite database as it is possible that other processes are using it, which could lead to possible corruption.
> 
> Instead it is recommended to use the sqlite `.backup FILE` or `.save FILE` functions. This can be invoked from the command line like `>sqlite3 my_db.sqlite ".backup backup_db.sqlite"`

There will always be the original unused copy created on compilation.

**Comment by KrisWilliamson** [comment]
I am getting the following error while running a MPI job on a server by myself (in my own namespace).
```
2024-01-19 19:45:42.733 [0] : database disk image is malformed
2024-01-19 19:45:42.912 [0] DB error: database disk image is malformed
```
The thing is, the sqqlite database does not appear to be corrupted.

A `PRAGMA integrity_check;` return nothing, and `VACUUM`ing the database succeeds while it should fail on if it were actually corrupted.

Given that we are not getting a `Database locked` error, I am guessing that the issue is something other than sqlite, maybe a file lock that sqlite cannot deal with, possibly related to file access over the network.

**Comment by KrisWilliamson** [comment]
Branch https://github.com/StatCan/openmpp/tree/31-model-data-isolation created

**Comment by vexingly** [comment]
Do you know what specific component/command is throwing the "database disk image is malformed" error message? Finding the specific cause of this is probably a good step to take.

Simplest method to test if the concurrent user access is the issue, would be to give each notebook their own separate model folder within the buckets folder. Need to keep the models in the blob storage as the notebook storage is too expensive to provision more for each user.

Making copies of the databases could be expensive (in disk/time) if they get too large.

Are there options for an sqlite database to make it more friendly to concurrency? Can we implicitly check the integrity/fix the database before starting a run?

A future option might be to explore other databases that are more resilient than sqlite.

**Comment by Souheil-Yazji** [comment]
All cases set to 1000

**UNIQUE Constraint Violation:** for this run, I had 6 subs running in parallel on 6 workers (1 thread each), taking about 0.9s to complete. Only the last worker encountered this error when writing to db

```
2024-01-26 19:39:17.049 [0] : UNIQUE constraint failed: All_Cancer_Cases_After_Intervention_a26ada28c.run_id, All_Cancer_Cases_After_Intervention_a26ada28c.acc_id, All_Cancer_Cases_After_Intervention_a26ada28c.sub_id, All_Cancer_Cases_After_Intervention_a26ada28c.dim0, All_Cancer_Cases_After_Intervention_a26ada28c.dim1, All_Cancer_Cases_After_Intervention_a26ada28c.dim2, All_Cancer_Cases_After_Intervention_a26ada28c.dim3
2024-01-26 19:39:23.981 [0] DB error: UNIQUE constraint failed: All_Cancer_Cases_After_Intervention_a26ada28c.run_id, All_Cancer_Cases_After_Intervention_a26ada28c.acc_id, All_Cancer_Cases_After_Intervention_a26ada28c.sub_id, All_Cancer_Cases_After_Intervention_a26ada28c.dim0, All_Cancer_Cases_After_Intervention_a26ada28c.dim1, All_Cancer_Cases_After_Intervention_a26ada28c.dim2, All_Cancer_Cases_After_Intervention_a26ada28c.dim3

```


**DB error: Disk I/O error:** here, I had 4 subs running on 5 single threaded workers (I know it doesn’t make sense, this is purely exploratory), the 0th worker didn’t run the sim and returned the error. Safe to assume it’s due to the distribution of work?
```
2024-01-26 18:55:12.305 [4] member=3 Write output tables - start
2024-01-26 18:55:12.314 [1] End Age standardize all cancers.
2024-01-26 18:55:12.314 [1] member=0 Write output tables - start
2024-01-26 18:55:12.316 [2] End Age standardize all cancers.
2024-01-26 18:55:12.316 [2] member=1 Write output tables - start
2024-01-26 18:55:12.320 [4] member=3 Write output tables - finish
2024-01-26 18:55:12.328 [1] member=0 Write output tables - finish
2024-01-26 18:55:12.330 [2] member=1 Write output tables - finish
2024-01-26 18:55:12.394 [3] End Age standardize all cancers.
2024-01-26 18:55:12.394 [3] member=2 Write output tables - start
2024-01-26 18:55:12.409 [3] member=2 Write output tables - finish
2024-01-26 18:57:37.321 [0] : disk I/O error
2024-01-26 18:57:43.604 [0] DB error: disk I/O error
```

**Database disk image malformed:** This one is the most interesting, 4 subs for 2 workers with 2 threads. Worker 1 ran members 0,1,2 (then terminated due to worker 0 kill signal), worker 0 did not run anything and produced the error. 
```
2024-01-26 18:38:20.901 [1] member=2 Prepare for simulation
2024-01-26 18:38:20.913 [0] : database disk image is malformed
2024-01-26 18:38:20.929 [0] DB error: database disk image is malformed
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD
with errorcode 1.
NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
2024-01-26 18:38:20.997 [1] member=2 Simulation progress=0% cases=0
2024-01-26 18:38:21.145 [1] member=2 Simulation progress=10% cases=25
2024-01-26 18:38:21.310 [1] member=2 Simulation progress=20% cases=50
2024-01-26 18:38:21.460 [1] member=2 Simulation progress=30% cases=75
2024-01-26 18:38:21.600 [1] member=2 Simulation progress=40% cases=100
2024-01-26 18:38:21.766 [1] member=2 Simulation progress=50% cases=125
2024-01-26 18:38:21.915 [1] member=2 Simulation progress=60% cases=150
```

**Comment by KrisWilliamson** [comment]
Another thing to try.
Under Advanced Run Options, there are several options that could reduce database access.
- Log Progress Percent - defaults to one, so write an entry to the log every percent.  change to 5% 
- Log Progress Step - similar to above, reduce I/O
- Sparse Output Tables - Do not store small values and zeros - reduce I/O and data stored in the db.

**Comment by Souheil-Yazji** [comment]
> Another thing to try. Under Advanced Run Options, there are several options that could reduce database access.
> 
> * Log Progress Percent - defaults to one, so write an entry to the log every percent.  change to 5%
> * Log Progress Step - similar to above, reduce I/O
> * Sparse Output Tables - Do not store small values and zeros - reduce I/O and data stored in the db.

I've actually been setting progress step to 0 and progress percent to 10 and tried with and without sparse tables. We can grow this thread a very good bit :)

**Comment by KrisWilliamson** [comment]
Source of error messages:
- Failed to write output table : /main/openm/libopenm/models/modelBase.cpp
- db transaction is already active : /main/openm/libopenm/db/dbExecSqlite.cpp
- DB error : multiple places including built into the model executable
  - also: /main/openm/omc/omc.cpp
- disk I/O error : /main/openm/libsqlite/sqlite3.c
- database disk image is malformed : /main/openm/libsqlite/sqlite3.c

**Comment by KrisWilliamson** [comment]
OpenM Database connection string
You can specify database connection string as command line argument, i.e.:
`modelOne.exe -OpenM.Database "Database=C:\My Model\m1.sqlite; Timeout=86400; OpenMode=ReadWrite;"`

a PostgreSQL connection string
`Driver={PostgreSQL};Server=IP address;Port=5432;Database=myDataBase;Uid=myUsername;Pwd=myPassword;`

It should be simple to manually test if this works

There is a `create_db.sql` file in the `main/sql` directory that creates the required database structure, and another file `insert_default.sql` that populates these tables with the required entries.

It is currently unknown if and how these scripts are invoked by OpenM.

**Comment by KrisWilliamson** [comment]
The common part of the makefile makefile.models.mak appears to be hardcoded to use a SQLite database.  It is not obvious how to specify another database to use.
